### ISSUE TITLE
[HOLD #11768] mWeb / Android - Popovers aren't closing on navigating back in native - reported by @thesahindia

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#9823]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/docs/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

### Back button

1. Go to any chat
2. Open emoji picker
3. Press native back button (back button of your device)

### Swipe back

1. Open any chat that has a message
2. Long press on a message
4. Without letting your finger leave the screen during the long press, swipe right
5. The app goes to the previous screen but context menu stays active

## Expected Result:
Popover should close

## Actual Result:
Popover don't close 

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Mobile Web

**Version Number:** 1.1.79-0
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://user-images.githubusercontent.com/43995119/178182147-8c0dcb4b-a013-4ea5-ab31-d24eb9da3455.mp4

https://user-images.githubusercontent.com/43996225/214363094-a18a9cb7-fdd6-4ae4-a7a2-4aed3ff06b21.mov

**Upwork job URL:** https://www.upwork.com/jobs/~01402faaf5968f784d
**Issue reported by:** @thesahindia
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1656497620210449

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by deetergp** [comment]
Definitely a contributor-worthy issue!

**Comment by michaelhaxhiu** [comment]
Job posted to Upwork - https://www.upwork.com/jobs/~010c2cbc9e40b3a07a

**Comment by michaelhaxhiu** [comment]
I don't have an android device to test it myself, but let's ask @mvtglobally if she has a moment to do so!

**Comment by michaelhaxhiu** [comment]
btw @jeet-dhandha were you unable to reproduce it yourself when you tried most recently?

**Comment by jeet-dhandha** [comment]
@michaelhaxhiu Just tried few minutes ago not able to reproduce it. If you have a macbook then you can try it on you android emulator too. Please tell me if you need steps for that.

**Comment by ahmdshrif** [comment]
@michaelhaxhiu I can repro it on production . ( mWeb )

https://user-images.githubusercontent.com/21364901/179274346-30c58e40-d6d6-408c-8bd5-05c95cc0d4ae.mp4

**Comment by nimamt** [comment]
**Proposal**
Handle this problem for the web platform (we don't see this bug on the android native platform/apk/google play) using the library _react-navigation_.

The final behavior of the web version is **identical** to the android native version, and fixes the bug for all popovers.

**NavigationRoot.js**
```JSX
if(Platform.OS === 'web') {
    if(this.props.modal?.willAlertModalBecomeVisible && this.lastState.current) {
        Navigation.navigationRef.reset(this.lastState.current);
        const event = new Event('close-modal');
        document.dispatchEvent(event);
        return;
    }
    this.lastState.current = state;
}
```
If navigation occurs when a modal is open (wrong behavior), revert that navigation and send an event to close the modal.
Correct behavior: The modal should be closed, and no navigation should occur. The next navigation may be accepted, if a model is not visible.

**NOTE:** https://reactnavigation.org/docs/preventing-going-back/
A similar solution to this would not work, I've tried it, with variations and listening for other events.

**BaseModal.js**
```JSX
this.handleShouldClose = this.handleShouldClose.bind(this);
document.addEventListener('close-modal', this.handleShouldClose);
```
Add the code for handing the _modal-close_ event to the constructor.

```JSX
handleShouldClose() {
    if(this.props.isVisible) {
        this.props.onClose();
        return true;
    }
    return false;
}
```
**Code of the event handler**. More than one modal exist at the same time, but only **one modal** is visible at any given time. For example out of 10 modals, 1 is visible and 9 others are not visible. This is not my own assumption, I picked this up from the code and comments for the modal object properties (willAlertModalBecomeVisible , isVisible) in Onyx.
Return false: Did not handle, some other handler should handle this
Return true: This event has been handled.

**Expensify.js**
```JSX
Onyx.merge(ONYXKEYS.MODAL, {willAlertModalBecomeVisible: false});
```
Add this to the constructor.
willAlertModalBecomeVisible is correct, but there is a bug where the app may be closed while a modal is open. When the app is opened again, Onyx keeps the data and thinks that `willAlertModalBecomeVisible:true`, while it should be reset to `false`, as there is no modal open initially.
**The property isVisible in the codebase is not correct. (bug??)

**

**Root Cause**
The back button does not cause an event on the web platform, it only causes a change in navigation.
https://stackoverflow.com/questions/57764033/custom-event-for-android-back-button-in-pwa
The only way to deal with this on any web framework is to use navigation. (Android back button/back or forward in browser all cause changes in navigation)

**Attachments**
**Demonstration**: A video demonstrating that the bug has been fixed.
https://user-images.githubusercontent.com/56079110/180610788-bd564e73-1747-4d0b-b9c9-1b454ac8697b.mp4
**Explanation.zip**: A zip containing a video explaining the code, watch this after reading all of the above text.
[Explanation.zip](https://github.com/Expensify/App/files/9174024/Explanation.zip)
**JS.zip**: The updated files.
[JS.zip](https://github.com/Expensify/App/files/9174022/JS.zip)

**Comment by michaelhaxhiu** [comment]
Updating job bounty to $500

**Comment by iwiznia** [comment]
Still waiting for proopsals mevlin. But also @parasharrajat did you see the proposal above?

**Comment by parasharrajat** [comment]
Checking on it. Still need to review a couple of things. I am not sure if we want to take that path.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by michaelhaxhiu** [comment]
Interesting - so this doesn't need assignment it seems! Definitely holding off on doubling the price any more.

**Comment by nimamt** [comment]
Mates, I don't believe it has been fixed. I've checked staging and it is still reproducible. And I have reproduced it by visiting both the staging site and compiling the code before.

Additionally, I would like to make an argument about one part of the philosophy section of the Readme. In its 5.ii section (Cross Platform 99%) it is said that  "If there is a feature that simply doesn't exist on all platforms and thus doesn't exist in RN". This is obviously NOT true. Check the below links, and it has been explicitly stated that these features are not available on all platforms, and yet they both exist in RN.

https://reactnative.dev/docs/modal#onrequestclose
https://reactnative.dev/docs/backhandler

Now if you check the source code for react-native-modal, it uses both of the aforementioned features of react-native. The android back button is simply not handled in a web version. This back button would only cause backward navigation (similar to the back button on your browser). Consequently, to deal with the modals remaining visible on the web version you would have to do something when navigating somewhere else, with the even simpler and minimal solution being closing the modals.  

https://snack.expo.dev/@nimamt/a411aa

You can see for yourself by testing the above code on your Android phone's browser, and the Android app version.

**Comment by nimamt** [comment]
# Proposal 2
We should use an invisible helper screen to also reflect the popovers/modals in `react-navigation`, as we are using `react-native-modal` and not following the recommended way of `react-navigation` docs.
**!!!IMPORTANT** See this: https://reactnavigation.org/docs/modal
Note that the recommended way to open a modal is to use a screen in react-native, that is to go to another screen when a modal is opened. And this is also true for popovers/modals that are not full-screen! See https://stackoverflow.com/questions/48992961/react-navigation-modal-height  
`React-navigation` should navigate to a screen/route called "Popover" when a `BaseModal `(`Popover` essentially) has been opened.
When we leave the "Popover" screen by pressing the back and forward browser buttons or the android back button or any other way that would be handled by `react-navigation`, the `react-navigation-modal` modal is also closed.
When the back button is handled on native versions of the app or the backdrop is pressed or any other way that would not close the "Popover" screen, we ensure that the screen would also be closed.
Again, the result is **identical** to the native version. Also, It is based on the **recommended** way to handle modals by `react-navigation`, and it is **cross-platform**.

## Code

### Navigation.js
```JSX 
function handleBaseModalOpen() {
    navigationRef.current.dispatch(StackActions.push("Popover"));
}

function handleBaseModalClosed() {
    const routes = navigationRef.current.getState().routes;
    const lastRouteName = routes[routes.length-1].name;
    if(lastRouteName === 'Popover')
        navigationRef.current.dispatch(StackActions.pop());
}
```
As mentioned before, we use an invisible "Popup" screen on top of our existing screens when a modal is opened.  
When a modal is closed by `react-native-modal` listeners, our screen in `react-navigation` is also closed to reflect this.
(And when the `react-navigation` "Popover" screen is closed, the `react-native-modal` modal is also closed to reflect this)  

### BaseModal.js
```JSX
constructor(props) {
    super(props);

    this.handleShouldClose = this.handleShouldClose.bind(this);
    **this.internalOnClose = this.internalOnClose.bind(this);**
    this.hideModal = this.hideModal.bind(this);
    document.addEventListener('close-modal', this.handleShouldClose);
}

internalOnClose() {
    this.props.onClose();
    Navigation.handleBaseModalClosed();
} 
```
If the modal is closed (internally) by `react-native-modal`, then the app must ensure the relative screen would be closed, other than having to call `this.props.onClose()` 

### AuthScreens.js
```JSX
class ModalScreen extends React.Component {

  render() {
    return (
      <View style={{ flex: 1 ,flexDirection: 'column', justifyContent: 'flex-end'}}>
      </View>
    );
  }
}

const baseModalHelperListeners = {
    beforeRemove: () => {
        const event = new Event('close-modal');
        document.dispatchEvent(event);
    },
};

<RootStack.Screen
    name="Popover"
    options={modalScreenOptions}
    component={ModalScreen}
    listeners={baseModalHelperListeners}
/>
```
The code for the "Popover" screen. Read the first part if you skipped it and you are clueless.
Regarding the listeners, this part is because when the `react-navigation` "Popover" screen is closed, the `react-native-modal` modal should also be closed to reflect this.

## Attachments
As I've said the behavior is basically identical to the native version, also I've provided a video of this correct behavior in a previous proposal.

All the edited js files are in this zip.
**Github Servers:** [Files.zip](https://github.com/Expensify/App/files/9273987/Files.zip)
**Mirror:** https://www.mediafire.com/file/bs06dy4r64tsmxa/Files.zip/file

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by michaelhaxhiu** [comment]
Hey @parasharrajat any update on this one? Should we double price? 

Also @mvtglobally why can't you reproduce it and others claim they can?

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

@michaelhaxhiu Not sure why this is not consistently repro, maybe it is device specific or smth. We have different testers running validations each time, so devices might differ

**Comment by iwiznia** [comment]
@michaelhaxhiu let's double this?

**Comment by iwiznia** [comment]
@parasharrajat did you review the proposals?

**Comment by nimamt** [comment]
"Definitely a contributor-worthy issue!"
"This is not yet fixed. It would be relatively complex to achieve but I am open to ideas."
"We already closed a similar issue in the past due to a complex solution."
Well, you did know it was complex from the beginning, why waste my time? :-)

**Comment by nimamt** [comment]
Besides, I've also given you a simple solution: Close the modal when navigation happens

**Comment by iwiznia** [comment]
@parasharrajat what's the status of this?

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Third week)

**Comment by iwiznia** [comment]
Bump again @parasharrajat

**Comment by parasharrajat** [comment]
No update. Unplanned leaves last week. There is an open proposal that I need help finalizing. I will discuss this internally. (will post a link if it is slack discussion)

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Forth week) 
We are unable to repro consistently.

**Comment by parasharrajat** [comment]
I will check it today and see if I can still reproduce this. I will also post a video for the same, if reprducible.

**Comment by michaelhaxhiu** [comment]
Perfect, will keep an eye for the reproduction attempt. If you can't reproduce let's close it.

**Comment by michaelhaxhiu** [comment]
@parasharrajat were you able to reproduce? If not, should we request a hand in #expensify-open-source to do with an Android?

**Comment by michaelhaxhiu** [comment]
Going to decide today if we (A) can't reproduce & close this GH or (B) keep open & double it.

**Comment by michaelhaxhiu** [comment]
Thanks @thesahindia. Doubled.

**Comment by 1notchdev** [comment]
Do you need more proposals for this issue yet?

**Comment by parasharrajat** [comment]
Happy to hear if the proposals are different from older ones. For now, working on a high-priority issue.

**Comment by michaelhaxhiu** [comment]
👋 Appreciate the polite approach of asking @1notchdev - in the future, if you see the `help wanted` label on a GH in this repo, that means it's available for hire and we are accepting proposals. 

@parasharrajat mentions a savvy tip about ensuring your proposal is different from others provided (mainly to save you time from creating a potential proposal that's already shared).

**Comment by michaelhaxhiu** [comment]
Price doubled!

**Comment by oogushikun** [comment]
@parasharrajat  I believe this bug is related to the comment you left on this issue https://github.com/Expensify/App/issues/7010#issuecomment-1018194823 referencing another issue on react-native-web https://github.com/necolas/react-native-web/pull/1738. also this is affecting all the modals, not just the emojipicker.

**Comment by parasharrajat** [comment]
No update.

**Comment by michaelhaxhiu** [comment]
I am requesting a buddy check in #expensify-open-source **[here](https://expensify.slack.com/archives/C01GTK53T8Q/p1668021625388299?thread_ts=1667941166.176039&cid=C01GTK53T8Q)**. @parasharrajat you may be able to unblock my question 🤓

**Comment by parasharrajat** [comment]
In discussions on the Slack. Close to conclusion.

**Comment by iwiznia** [comment]
What is the discussion? The one linked by @michaelhaxhiu ?

**Comment by parasharrajat** [comment]
Based on the latest discussion https://expensify.slack.com/archives/C01GTK53T8Q/p1668198693474999?thread_ts=1668027443.464419&cid=C01GTK53T8Q, we decided to do close the modal/popovers as well as navigating the user to the previous page in history.

This will keep the default browser behavior and fix the issue of stuck popovers/modals. 


@michaelhaxhiu Can you please update the expected behaviour on the description?

**Comment by parasharrajat** [comment]
This issue is now taking proposals actively. Please reshare your proposals based on the new expected behavior https://github.com/Expensify/App/issues/9823#issuecomment-1317686264 if you already proposed.

**Comment by kevinksullivan** [comment]
Nice conclusion here! @michaelhaxhiu I figure we can move this forward with a fresh post to #expensify-open-source with the updated issue?

**Comment by marcaaron** [comment]
Looked into this and pretty sure this is actually related to the `Freeze` component used on the `ReportScreen`. When we go back using the browser while a modal is up it gets stuck since the modal visibility is dependent on state. since the component is frozen (which stops it from re-rendering) the modal never closes.

This fixes the issue:

```diff
diff --git a/src/pages/home/ReportScreen.js b/src/pages/home/ReportScreen.js
index 972dc0210f..b6d69bc234 100644
--- a/src/pages/home/ReportScreen.js
+++ b/src/pages/home/ReportScreen.js
@@ -218,7 +218,7 @@ class ReportScreen extends React.Component {
         const isLoadingInitialReportActions = _.isEmpty(this.props.reportActions) && this.props.report.isLoadingReportActions;
         return (
             <Freeze
-                freeze={this.props.isSmallScreenWidth && this.props.isDrawerOpen}
+                freeze={false}
                 placeholder={(
                     <ScreenWrapper
                         style={screenWrapperStyle}
```

However, disabling the Freeze component will re-introduce a bug where the previous screen shows briefly before the next screen when switching chats.

We are gonna get rid of the Freeze components as soon as we improve the react navigation stuff (it's kind of a hack IMO). But if we do it now this is what we'll see on native:

https://user-images.githubusercontent.com/32969087/202563174-c59f05ec-db92-4ddc-ba99-1a5d92be360f.mp4

**Comment by JmillsExpensify** [comment]
@iwiznia To help clear out the extensive backlog of /App bugs, we're putting the spotlight on all bugs older than 4 weeks old. To help unblock the roadmap and get our bug pipeline back in equilibrium, can you:

 - Decide whether any proposals currently meet our guidelines and can be approved as-is
 - For any that can't, please take this issue internal and treat it as one of your highest priorities
 - If you have any questions, don't hesitate to start a discussion in #bug-zero

Thanks everyone!

**Comment by JmillsExpensify** [comment]
P.S. I equally valid solution is to link this issue to react navigation issue and then put this issue on hold until then.

**Comment by marcaaron** [comment]
I think this issue is really not so bad and could just wait. Worst case scenario you tap once more to close the menu. Normally, I think people would do the tap on mobile web to exit the modal before going back - so it's a little edge case. Let's pull it into the navigation stuff.

**Comment by JmillsExpensify** [comment]
Nice, I agree with that. I'll proactively do that and also assign myself now since I'm curating the tracking issue.

**Comment by iwiznia** [comment]
Not overdue, this is on hold

**Comment by JmillsExpensify** [comment]
Still on hold for navigation reboot.

**Comment by parasharrajat** [comment]
This is an internal issue and also on HOLD. We are not accepting proposals for it.

**Comment by JmillsExpensify** [comment]
Still held on the larger navigation re-write.

**Comment by JmillsExpensify** [comment]
Yep, thanks Melvin. Updated to monthly so you leave us alone for a bit.

**Comment by JmillsExpensify** [comment]
Still on hold for navigation.

**Comment by JmillsExpensify** [comment]
Still on hold for navigation. No ETA.

**Comment by JmillsExpensify** [comment]
Same same

**Comment by JmillsExpensify** [comment]
Still on hold

**Comment by JmillsExpensify** [comment]
Coming off hold very soon!

**Comment by iwiznia** [comment]
Going on sabbatical, please reassign another engineer if this goes off hold

**Comment by JmillsExpensify** [comment]
You're fine. I still need to get around to testing.

**Comment by JmillsExpensify** [comment]
Hold

**Comment by JmillsExpensify** [comment]
@thesahindia Would you mind re-testing this issue that you reported and confirm if it's still happening?

**Comment by JmillsExpensify** [comment]
Perfect, thank you! $250 payment due to @thesahindia for issue reporting.

**Comment by JmillsExpensify** [comment]
Given that you are on NewDot payments, I'm going to approve and close this issue.

**Comment by JmillsExpensify** [comment]
$250 payment approved for @thesahindia based on summary above.
